### PR TITLE
Flush triggered by DB write buffer size picks the oldest unflushed CF

### PR DESCRIPTION
**Pull request by siying**
Summary: Previously, when DB write buffer size triggers, we always pick the CF with most data in its memtable to flush. This approach can minimize total flush happens. Change the behavior to always pick the oldest unflushed CF, which makes it the same behavior when max_total_wal_size hits. This approach will minimize size used by max_total_wal_size.

Test Plan: Modify existing unit tests

### CODE REVIEW

File: db/db_impl.cc [hunk 1]
```diff
@@ -4759,18 +4759,18 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
       if (!cfd->mem()->IsEmpty()) {
         // We only consider active mem table, hoping immutable memtable is
         // already in the process of flushing.
-        size_t cfd_size = cfd->mem()->ApproximateMemoryUsage();
-        if (largest_cfd == nullptr || cfd_size > largest_cfd_size) {
-          largest_cfd = cfd;
-          largest_cfd_size = cfd_size;
+        uint64_t o = cfd->OldestLogToKeep();
```
Comment (al13n321) [inline db/db_impl.cc]: This won't work well if all writes have WAL disabled, right? OldestLogToKeep() will have the same value in all CFs, and this loop will pick up the oldest CF; correct?

Would it make sense to pick CF with the smallest (nonzero) GetFirstSequenceNumber() instead? Seems that it would have the same behavior as OldestLogToKeep() when WAL is used, but will also work when WAL is not used.

Comment (siying) [inline db/db_impl.cc]: Ahh, correct. Let me fix it.

**Comment by siying** [comment]
@al13n321 fixed it.

File: db/db_impl.cc [hunk 2]
```diff
@@ -4759,18 +4759,18 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
       if (!cfd->mem()->IsEmpty()) {
         // We only consider active mem table, hoping immutable memtable is
         // already in the process of flushing.
-        size_t cfd_size = cfd->mem()->ApproximateMemoryUsage();
-        if (largest_cfd == nullptr || cfd_size > largest_cfd_size) {
-          largest_cfd = cfd;
-          largest_cfd_size = cfd_size;
+        uint64_t seq = cfd->mem()->GetLatestSeqWhenCreate();
```
Comment (al13n321) [inline db/db_impl.cc]: GetLatestSeqWhenCreate() effectively tells the time of last flush in this memtable, right? That's probably not what we want: in our use case most memtables are empty most of the time, and the time of last flush in a CF seems quite irrelevant. We want to flush the oldest unflushed data; the oldest unflushed piece of data lives in memtable with the smallest GetFirstSequenceNumber(), right?

Comment (siying) [inline db/db_impl.cc]: Empty memtables are skipped because of line 4759.

Comment (al13n321) [inline db/db_impl.cc]: Do you see any advantages of using GetLatestSeqWhenCreate() instead of GetFirstSequenceNumber() here? I can't think of any. Some advantages of GetFirstSequenceNumber():
 * Has a property that's important for logdevice: suppose you wrote a key-value pair A, then wrote db_write_buffer_size bytes of data to the DB (any CFs). Write A is (almost) guaranteed to be have been flushed by now. Imo this also makes it easy to reason about.
 * WAL never grows above 2*db_write_buffer_size. (Of course there's max_total_wal_size, but in logdevice we're going to stop using WAL and have our own external WAL-like thing that we don't want to grow too much; max_total_wal_size won't help with that.)
 * Less code.

A (fairly realistic with our workload) scenario in which GetLatestSeqWhenCreate() fails the first two advantages:
 * db_write_buffer_size = 1 GB. There are 1000 CFs. No one writes to them atm.
 * A new CF is created, a bunch of stuff is written to it (some flushes happen), then writes to it stop. A small amount of data is still left in its memtable.
 * 900 MB are written to CF 1, then 900 MB are written to CF 2 and so on up to CF 1000. During this whole time the memtable in CF 1001 stays unflushed. WAL grows to 900 GB.

Comment (al13n321) [inline db/db_impl.cc]: Re empty memtables: of course I didn't think that this code would flush empty memtables; my problem is with the order of flushing nonempty memtables.

Comment (siying) [inline db/db_impl.cc]: If we want to make sure we always effectively enforce we try the flush the memtable to free the log files, GetLatestSeqWhenCreate() is safer, because it is closer to how we determine whether a log file can be dropped. We only track the log files to be kept while we switch the memtable and never tracked which log file the first key of the memtable goes to. If a new memtable for CF1 is crated while with log file 00003.log, but the first key to CF1 is not inserted until 00004.log, 000003 is still being held util CF1 is flushed. If you want us to guarantee 000003.log is flushed as soon as possible, a switch sequence ID is a safer way. If you are interested in reading the code, here is where you can start: https://github.com/facebook/rocksdb/blob/5.3.fb/db/db_impl.cc#L766 It's pretty compliated though.

Also, GetFirstSequenceNumber() is a complicated variable. It is updated using a CAS logic and I searched the code and no code is relying on the value of it (except 0 or 1) in critical logic. I'm wary of depending on a complicated logic.

I hope we can keep the implementation simply so it is less prone to unexpected behavior. Would this approach generally satisfy your requirement?

Comment (al13n321) [inline db/db_impl.cc]: > it is closer to how we determine whether a log file can be dropped

The current GetLatestSeqWhenCreate() is missing one important aspect of how we determine whether a log file can be dropped: https://github.com/facebook/rocksdb/blob/5.3.fb/db/db_impl.cc#L5287-L5294

That's my only concern about GetLatestSeqWhenCreate(). It's a major concern for us, as I explained with an example. The straightforward way to fix it is to update latest_seq_when_create_ in SwitchMemtable() for empty CFs, similar to how SetLogNumber() is done; except it has to be done even if creating_new_log is false, to make it work with our walless setup.

Fwiw, if latest_seq_when_create_ is updated in SwitchMemtable() for empty CFs, I don't see any difference between using GetFirstSequenceNumber() vs GetLatestSeqWhenCreate() (vs DBImpl::MinLogNumberToKeep() too, if WAL is enabled and 2pc is disabled). Can you point out any differences?

If you're more comfortable with GetLatestSeqWhenCreate() than GetFirstSequenceNumber(), I don't mind, but we need that "optimization" in SwitchMemtable().

**Comment by siying** [comment]
Address @al13n321 comment. Make the logic map the log file dropping criteria.

File: db/memtable.h [hunk 3]
```diff
@@ -286,6 +286,15 @@ class MemTable {
     return earliest_seqno_.load(std::memory_order_relaxed);
   }
 
+  // DB's latest sequence ID when the memtable is created.
```
Comment (al13n321) [inline db/memtable.h]: Not true anymore.

**Comment by siying** [comment]
The valgrind failure is not related.
